src/ssl.c

```diff
@@ -15924,198 +15943,150 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
-            node->num  = (list == NULL) ? 1 : list->num + 1;
-            node->next = list;
-            list = node;
+                nameCopy->x509 = NULL;
 
-            wolfSSL_X509_free(cert);
-            cert = NULL;
-        }
-
-        err = wolfSSL_ERR_peek_last_error();
-
-        if (ERR_GET_LIB(err) == ERR_LIB_PEM &&
-                ERR_GET_REASON(err) == PEM_R_NO_START_LINE) {
-            /*
-             * wolfSSL_PEM_read_bio_X509 pushes an ASN_NO_PEM_HEADER error
-             * to the error queue on file end. This should not be left
-             * for the caller to find so we clear the last error.
-             */
-            wc_RemoveErrorNode(-1);
-        }
+                if (wolfSSL_sk_X509_NAME_push(list, nameCopy) !=
+                        WOLFSSL_SUCCESS) {
+                    WOLFSSL_MSG("wolfSSL_sk_X509_NAME_push error");
+                    /* Do free in loop because nameCopy is now responsibility
+                     * of list to free and adding jumps to cleanup after this
+                     * might result in a double free. */
+                    wolfSSL_X509_NAME_free(nameCopy);
+                    goto cleanup;
+                }
 
-        wolfSSL_X509_free(cert);
-        wolfSSL_BIO_free(bio);
-        return list;
-    #else
-        (void)fname;
-        return NULL;
-    #endif
-    }
-    #endif
-#endif /* !NO_BIO */
-#endif /* OPENSSL_EXTRA || HAVE_WEBSERVER */
+                wolfSSL_X509_free(cert);
+                cert = NULL;
+            }
 
-#ifdef OPENSSL_EXTRA
-    #if !defined(NO_RSA) && !defined(NO_CERTS)
-    int wolfSSL_CTX_add_client_CA(WOLFSSL_CTX* ctx, WOLFSSL_X509* x509)
-    {
-        WOLFSSL_STACK *node = NULL;
-        WOLFSSL_X509_NAME *subjectName = NULL;
+            err = wolfSSL_ERR_peek_last_error();
 
-        WOLFSSL_ENTER("wolfSSL_CTX_add_client_CA");
+            if (ERR_GET_LIB(err) == ERR_LIB_PEM &&
+                    ERR_GET_REASON(err) == PEM_R_NO_START_LINE) {
+                /*
+                 * wolfSSL_PEM_read_bio_X509 pushes an ASN_NO_PEM_HEADER error
+                 * to the error queue on file end. This should not be left
+                 * for the caller to find so we clear the last error.
+                 */
+                wc_RemoveErrorNode(-1);
+            }
 
-        if (ctx == NULL || x509 == NULL){
-            WOLFSSL_MSG("Bad argument");
-            return SSL_FAILURE;
+            err = WOLFSSL_SUCCESS;
+cleanup:
+            wolfSSL_X509_free(cert);
+            wolfSSL_BIO_free(bio);
+            if (err != WOLFSSL_SUCCESS) {
+                /* We failed so return NULL */
+                wolfSSL_sk_X509_NAME_pop_free(list, NULL);
+                list = NULL;
+            }
+            return list;
+        #else
+            (void)fname;
+            return NULL;
+        #endif
         }
+        #endif
+    #endif /* !NO_BIO */
+#endif /* OPENSSL_EXTRA || WOLFSSL_EXTRA || HAVE_WEBSERVER */
```

Comment:
Comment here is wrong. `HAVE_WEBSERVER` isn't part of the macro guard.

Reply:
Fixed.